src/Control/Lens/Traversal.hs

```diff
@@ -1466,3 +1468,28 @@ traverseByOf l pur app f = reifyApplicative pur app (l (ReflectedApplicative #.
 -- @
 sequenceByOf :: Traversal s t (f b) b -> (forall x. x -> f x) -> (forall x y. f (x -> y) -> f x -> f y) -> s -> f t
 sequenceByOf l pur app = reifyApplicative pur app (l ReflectedApplicative)
+
+-- | A version of 'Control.Lens.Setter.over' that forces the result(s) of
+-- applying the function. This can prevent space leaks when modifying lazy
+-- fields.
+--
+-- @
+-- over traverse (const ⊥) [1,2] = [⊥, ⊥]
+-- over' traverse (const ⊥) [1,2] = ⊥
+-- @
+--
+-- @
+-- over' :: 'Lens' s t a b -> (a -> b) -> s -> t
+-- over' :: 'Traversal' s t a b -> (a -> b) -> s -> t
+-- @
+over' :: LensLike Solo s t a b -> (a -> b) -> s -> t
+over' l f = getSolo . l (\old -> Solo $! f old)
```

Comment:
It maybe worth to add a quick note (or reference) why one needs a boxed type here, i.e. `Solo` and that using `Identity` won't work. (I actually don't know why, someone else looking how `lens` is implemented may wonder as well).

Secondly, we should document (probably publicly) that `Solo` can be given `Settable` instance, but probably shouldn't to prevent this combinator working with true setters and lazy functors (as they should be, I guess?)

```haskell
>>> length $ over' mapped (const undefined) [1,2]
2
```

(https://github.com/ekmett/distributive/issues/67 somewhat related).

Reply:
`traverse @Identity` is operationally identical to `fmap` (see, e.g., `fmapDefault`). A simple example:

```haskell
data Pair a = Pair a a
  deriving (..., Traversable)

traverse @Pair @Solo f (Pair a b)
= liftA2 Pair (f a) (f b)
= case (f a, f b) of
      (Solo a', Solo b') -> Solo (Pair a' b')

traverse @Pair @Identity f (Pair a b)
= case (f a, f b) of
     (Identity a', Identity b') -> Identity (Pair a' b')
```

In the `Identity` case, the pattern matches on `Identity` don't force anything (it's just a newtype), but in the `Solo` case, they force the applications to WHNF to reveal the `Solo` constructors. That forces whatever computation `f` chooses to perform before revealing the constructor; in our case, it forces the result inside to WHNF.

Reply:
Yes, a `Settable` instance for `Solo` sounds bad from this perspective. I don't know enough about `lens` to say if there's ever any reason to want that, but it seems unlikely. Should we ban the instance as a `TypeError`?

Reply:
Don't explain it to *me*, explain it to the (future) reader of the code. This is why asked for a reference or a note.

That said, it probably should be explained in `base`'s `Solo` docs, so that could be used as reference; but there is only a side-comment mention about "non-leaky traversals".

Reply:
I learned yesterday that this idea has been invented before (rather unsurprisingly); I've edited, expanded, and adapted a note about it in `optics` for inclusion here.